src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,309 @@
+			return err
+		}
+	}
+
+	if upgrading {
+		r.logger.Info("Continuing cluster image update", "cluster image", newImage)
+	}
+
+	if upgrading || rpContainer.Image != newImage {
+		podSpec := &sts.Spec.Template.Spec
+		if err := r.modifyPodImage(podSpec, newImage); err != nil {
+			return err
+		}
+
+		if err := r.partitionUpdateImage(ctx, sts, newImage); err != nil {
+			return err
+		}
+
+		// Update is complete for all pods (and all are ready). Set upgrading status to false.
+		if err := r.updateUpgradingStatus(ctx, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) updateUpgradingStatus(
+	ctx context.Context, upgrading bool,
+) error {
+	if !reflect.DeepEqual(upgrading, r.pandaCluster.Status.Upgrading) {
+		r.pandaCluster.Status.Upgrading = upgrading
+		if err := r.Status().Update(ctx, r.pandaCluster); err != nil {
+			r.logger.Error(err, "Failed to update RedpandaClusterStatus")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string,
+) error {
+	replicas := *sts.Spec.Replicas
+
+	// When a StatefulSet's partition number is set to `i`, only Pods with ordinal
+	// greater than or equal to `i` will be updated.
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#partitions
+	for ordinal := replicas - 1; ordinal >= 0; ordinal-- {
+		// Update() on statefulset has not been called yet in this run, however,
+		// this could be a retry call in which case we skip the current partition.
+		poderr := r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr == nil {
+			r.logger.Info("Pod already updated, skip", "ordinal", ordinal)
+			continue
+		}
+
+		// Continue only if error is due to Pod not ready, or unchanged image
+		// as an attempt to fix the Pod.
+		if !errors.Is(poderr, errContainerHasWrongImage) && !errors.Is(poderr, errPodNotReady) {
+			return poderr
+		}
+
+		// Before continuing to update the ith Pod, verify that the previously updated
+		// Pod (if any) has rejoined its groups after restarting, i.e., is ready for I/O.
+		if err := r.ensureRedpandaGroupsReady(sts, replicas, ordinal+1); err != nil {
+			return &NeedToReconcileError{RequeueAfter: requeueDuration,
+				Msg:	fmt.Sprintf("redpanda on pod (ordinal: %d) not ready", ordinal)}
+		}
+
+		if err := r.rollingUpdatePartition(ctx, sts, ordinal); err != nil {
+			return err
+		}
+
+		// Restarting the Pod takes enough time to warrant a requeue.
+		return &NeedToReconcileError{RequeueAfter: requeueDuration,
+			Msg:	fmt.Sprintf("wait for pod (ordinal: %d) to restart", ordinal)}
+	}
+
+	// Ensure 0th Pod is ready for I/O before completing the upgrade.
+	if err := r.ensureRedpandaGroupsReady(sts, replicas, 0); err != nil {
+		return &NeedToReconcileError{RequeueAfter: requeueDuration,
+			Msg:	fmt.Sprintf("redpanda on pod (ordinal: %d) not ready", 0)}
+	}
```

Comment:
Because after we update each pod in the loop, we requeue. If the pod becomes ready in the meantime we'd have no chance to call `ensureRedpandaGroupsReady` for it.